subprojects/gradle/kubernetes/src/main/kotlin/com/avito/utils/gradle/KubernetesCredentials.kt

```diff
@@ -1,18 +1,38 @@
 package com.avito.utils.gradle
 
 import com.avito.kotlin.dsl.getMandatoryStringProperty
+import com.avito.kotlin.dsl.getOptionalStringProperty
 import org.gradle.api.Project
 import java.io.Serializable
 
-data class KubernetesCredentials(
-    val token: String,
-    val caCertData: String,
-    val url: String
-) : Serializable
+sealed class KubernetesCredentials : Serializable {
+
+    data class Service(
+        val token: String,
+        val caCertData: String,
+        val url: String
+    ) : KubernetesCredentials(), Serializable
+
+    data class Config(
+        val context: String,
+        val caCertFile: String? = null,
+        val configFile: String = "${System.getProperty("user.home")}/.kube/config"
+    ) : KubernetesCredentials(), Serializable
+}
 
 val Project.kubernetesCredentials: KubernetesCredentials
-    get() = KubernetesCredentials(
-        token = this.getMandatoryStringProperty("kubernetesToken"),
-        caCertData = this.getMandatoryStringProperty("kubernetesCaCertData"),
-        url = this.getMandatoryStringProperty("kubernetesUrl")
-    )
+    get() {
+        val context = getOptionalStringProperty("kubernetesContext", nullIfBlank = true)
+        return if (context.isNullOrBlank()) {
+            //todo it should be mandatory only in avito ci
+            // move it to instrumentationPlugin config to control
+            KubernetesCredentials.Service(
+                token = getMandatoryStringProperty("kubernetesToken"),
+                caCertData = getMandatoryStringProperty("kubernetesCaCertData"),
+                url = getMandatoryStringProperty("kubernetesUrl")
+            )
+        } else {
+            val caCertFile = getOptionalStringProperty("kubernetesCaCertFile", nullIfBlank = true)
```

Comment:
Why is it optional? Is `kubernetesContext` not enough for separate logic?

Reply:
because custom certificate is optional, by default its data parsed from kube/config as plain bytes